src/libs/elektra/contracts.c

```diff
@@ -0,0 +1,124 @@
+/**
+ * @file
+ *
+ * @brief Contract constructors for kdbOpen()
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include "kdb.h"
+#include "kdbprivate.h"
+
+#include <stddef.h>
+
+/**
+ * Sets up a contract for use with kdbOpen() that configures
+ * the gopts plugin.
+ *
+ * @param contract    The KeySet into which the contract will be written.
+ * @param argc        The argc value that should be used by gopts.
+ * @param argv        The argv value that should be used by gopts.
+ *                    IMPORTANT: The pointer and data behind must be valid
+ *                    until after kdbClose() is called.
+ * @param envp        The envp value that should be used by gopts.
```

Comment:
You can just pass `environ`. That's why the function takes pointer arguments and stores them as binary keys.

Reply:
This was not the question: the question is if someone does not have the environ pointer (or argv or argc) and wants the gopts plugin to search for the environ like gopts does it when being mounted globally without contracts. Probably it is by passing NULL but this is not documented.

Reply:
Passing `NULL` is an error. Right now the `gopts` plugin either takes all values from the global keyset or none of them. We could allow overriding just `argc`/`argv` or just `envp`, but I'm not sure it's needed.

I would strongly recommend always passing `argc` and `argv` (or using `elektraGOptsContractFromStrings` if that's not possible) and never actually relying on the `gopts` internal command line argument reading. I would even go as far as entirely removing the old way of reading the arguments in a future release. IMO the code is just too system dependent and fragile. Especially, since the way `gopts` tries to read the arguments is determined at compile-time.

For environment variables it is a bit different, those can be easily accessed inside of `gopts`. But that is because the POSIX standard says that you can declare `extern char ** environ` literally anywhere and you will get a pointer to the environment variables. So anybody calling from C or C++ should be able to provide this Pointer without any problems.

With `elektraGOptsContractFromStrings` it's different. Here parsing is involved. So it could make sense to allow passing `NULL` for the environment variables to fallback to the `gopts` internal declaration of `environ` and avoid concatenating the environment into a single string and dissecting it again in `gopts`.

Reply:
> Passing NULL is an error. Right now the gopts plugin either takes all values from the global keyset or none of them. 

Please add this to the docu. I think elektraGOptsContract should somehow offer a similar functionality than what you can do when creating your own KeySet, so passing all 3 NULL should attempt to automatically figure out args+envs.

> I would strongly recommend always passing argc and argv

Yes, I think we actually should recommend not passing NULLs in the docu. But it is not always possible (e.g. if you implement a lib), so it is nice if such auto-detection code is available.

> With elektraGOptsContractFromStrings it's different. Here parsing is involved. So it could make sense to allow passing NULL for the environment variables to fallback to the gopts internal declaration of environ and avoid concatenating the environment into a single string and dissecting it again in gopts.

Yes, this is another reason.

Reply:
> passing all 3 NULL should attempt to automatically figure out args+envs.

Yes, passing `argc=0, argv=NULL, envp=NULL` should actually be a valid option. I'll change that and allow `NULL`s in `elektraGOptsContractFromStrings`.